### PR TITLE
Expand inputs and avoid duplicates

### LINKED ISSUE #758: Develop a way to expand `inputs` and avoid duplicates
When we request `get_asset_inputs_for_amount` from the node, we'll get resources of the node's choosing following a randomized algorithm to prevent dust accumulation.

During the formation of our tx we ask for resources to cover the gas fee and cover whatever assets the contract calls required. The node is free to return whichever resources it deems appropriate to cover our request.

Later on, after we've formed the tx, somebody might want to add some more inputs for whatever reason. They're probably going to do that using extend_inputs. And they're probably going to get those inputs by calling `get_asset_inputs_for_amount`.

There is nothing (AFAIK) preventing the possibility that the node is going to return resources that we've already 'reserved' by either covering the transaction costs or because a contract call demanded them. The node can't know we're planning to do anything with the resources it had suggested just a few moments before while we were building the tx.

So a user might find this snippet to work most of the time:

```rust
let mut a_call = contract.methods().some_method().get_executable_call();
let some_more_inputs = wallet.get_asset_inputs_for_amount(...);
a_call.extend_inputs(some_more_inputs);
```

but once in a while, the node will return a resource we've already placed in our tx inputs and then this code will error out due to the duplicated ids. This is problematic.

We could either:

1- leave the user to his own devices and have him request all his resources and filter out the ones we've already chosen for his tx, or repeat get_asset_inputs_for_amount until the node changes its mind, or,
2- develop a way of extending inputs such that we request resources from the node only once for a given asset id.
    both of the above, let the user use the error-prone one if he knows what he's doing

The `#1` way is more flexible because the user would choose exactly which resource he wants to extend as inputs. I'm out of my depth here but that could be used to make decisions in the contract? Heard about predicates making decisions based on inputs.

The  `#2` way is more friendly, just have the user extend the inputs by noting his wish of (asset_id, amount) and we'll find the coins for that wish and for our other needs at the same time, avoiding duplicates.
So something like:
```rust
call.add_input_resources(asset_id, amount);
```
and right before executing the script, we would do one pass per asset id to actually find all the needed resources.

### PR DESCRIPTION
**Pull request by Salka1988**
closes #758

```rust 
pub struct PrepareExecutableFuelCall {
    pub(crate) tx: fuels_core::tx::Script,
    pub(crate) calls: Calls,
    pub(crate) wallet: WalletUnlocked,
    pub(crate) inputs: Vec<Input>,
    pub(crate) outputs: Vec<Output>,
}
```
The `from_contract_calls` function enables us to convert a `ContractCall` into an executable call. We can add `inputs` and `outputs` without worrying about duplicates, as the `prepare` function will be called afterwards to collect all entries and perform the necessary checks for execution. The resulting `ExecutableFuelCall` structure should not be modified unless we are confident in our actions.

### CODE REVIEW

**Review by MujkicA — COMMENTED** [review]
I think this doesn't remove the original issue. Seems to me like the only options we have is to relinquish full responsibility for adding enough resources to cover fees to the user if they choose to modify inputs or only allow them to add certain amount/asset combinations without specifying the messages/coins that should constitute the inputs.

File: examples/cookbook/src/lib.rs [hunk 1]
```diff
@@ -218,7 +218,7 @@ mod tests {
         let contract_instance = MyContract::new(contract_id, wallet_1.clone());
 
         let call_handler = contract_instance.methods().initialize_counter(42);
-        let mut executable_call = call_handler.get_executable_call().await?;
+        let mut executable_call = call_handler.get_executable_call().await?.prepare().await?;
```
Comment (iqdecay) [inline examples/cookbook/src/lib.rs:221]: I don't like the UX of this, it is yet another chain of `method()` to just do a contract call.
From what I understand, the suggestion was to let the user do her thing, and then only do one pass for requesting resources (both for fees and resources) from the node.

File: packages/fuels-contract/src/contract_calls_utils.rs [hunk 2]
```diff
@@ -419,269 +382,6 @@ mod test {
         assert_eq!(custom_input, args[1].resolve(0));
     }
 
-    #[test]
-    fn contract_input_present() {
```
[the remark this answers is not in the document]
Comment (iqdecay) [reply packages/fuels-contract/src/contract_calls_utils.rs:423]: same question

**Review by iqdecay — CHANGES_REQUESTED** [review]
I'm not a big fan of the UX, and also agree with what Ahmed said.

File: packages/fuels-contract/src/contract_calls_utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Salka1988) [reply packages/fuels-contract/src/contract_calls_utils.rs:423]: these were the tests for the feature I deleted in this branch

**Comment by Salka1988** [comment]
After further discussion and consideration, we have decided to take a different approach to solving the problem, which you can be found on branch #778

**Pull request closed without merging**